Register SLM run before snapshotting to save stats

parkertimmins opened the pull request:
The SLM health indicator relies on the [policyMetadata.getInvocationsSinceLastSuccess](https://github.com/elastic/elasticsearch/blob/95275f5be625aee03a1917353cc3c255771f7a02/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorService.java#L219) to determine when the last several snapshots have failed.

If a snapshot fails and the master is shutdown before setting invocationsSinceLastSuccess, the fact that failur occurred will be lost. To solve this, before snapshotting, register in the cluster state that an slm snapshot is about to be run. If the run fails, and storing the failure information in the cluster state also fails due to a master shutdown, during the next snapshot run the fact that there was a failure can be determined because the failed run will be registered in the cluster state.

parkertimmins committed:
Register SLM run before snapshotting to save stats

Before snapshotting, store in the cluster state that
an slm snapshot is about to be run. If the run fails,
and storing the failure information in the cluster
state also fails, during the next snapshot run the fact
that there was a failure can be inferred from the presence
of an already registered run.

parkertimmins commented:
This PR is still needs some work, but I'd like to get some feedback to make sure it's going in the right direction!

parkertimmins commented on x-pack/plugin/slm/src/internalClusterTest/java/org/elasticsearch/xpack/slm/SLMStatDisruptionIT.java:
```diff
@@ -0,0 +1,550 @@
+        assertTrue(latch.await(1, TimeUnit.MINUTES));
+
+        // restart master so failure stat is lost
+        // TODO this relies on a race condition. The node restart must happen before stats are stored in cluster state, but this is not guaranteed.
```
There is a race condition here that I haven't figured how to remove. If the failure is recorded in the cluster state: https://github.com/elastic/elasticsearch/blob/ff7f00e553ccc0f9786a78d55b839b46b0e9ef03/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java#L131

before the master is restarted, the failure stat will not be lost. I'd like to test this without a race condition, but not sure how to do so.

parkertimmins commented on x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java:
```diff
@@ -204,6 +215,92 @@ public static String exceptionToString(Exception ex) {
+
+            SnapshotLifecyclePolicyMetadata.Builder newPolicyMetadata = SnapshotLifecyclePolicyMetadata.builder(policyMetadata);
+            if (preRegisteredRuns > numSnapshotsRunning) {
+                final long unrecordedFailures = preRegisteredRuns - numSnapshotsRunning;
```
Counting the number of currently running snapshots and using this to update preRegisteredRuns worries me a bit. A simpler alternative would be the following:  If there are concurrently running snapshots for the same policy, just don't update `invocationsSinceLastSuccess`. I _think_ that snapshots can only run concurrently due to a manual rather than scheduled run. It does not seem terrible to just bail out in this case and not get the benefits of this PR.

parkertimmins commented on x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java:
```diff
@@ -204,6 +215,92 @@ public static String exceptionToString(Exception ex) {
+                newPolicyMetadata.setInvocationsSinceLastSuccess(policyMetadata.getInvocationsSinceLastSuccess() + unrecordedFailures);
+
+                // There are likely scenarios where inc/decrements to preRegisteredRuns can be lost, so lower bound to 1 before run.
+                long newPreRegisteredRuns = Math.min(1, preRegisteredRuns + 1 - unrecordedFailures);
```
It seem there are likely situations where `preRegisteredRuns` and `numSnapshotsRunning` are inconsistent and thus `unrecordedFailures` gets a wrong value. Setting  `newPreRegisteredRuns` to a minimum value of 1 should mitigate this, but doesn't solve it.

One alternative to improve this would be to store a set of pre-registered snapshot ids rather than a single count. Then remove all snapshot ids that are currently running from the set. Use the size of the remaining set to increment invocationsSinceLastSuccess.

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

parkertimmins committed:
Update docs/changelog/110216.yaml

elasticsearchmachine commented:
Hi @parkertimmins, I've created a changelog YAML for you.

parkertimmins committed:
Fix license, spotless apply

parkertimmins committed:
previous spotty application of spotless apply

parkertimmins committed:
fix broken test

parkertimmins committed:
Start snapshot after registration stored in CS

Start snapshot after pre-registration cluster state update has succeeded.

parkertimmins committed:
Pre-register snapshot name rather than count

Store a set of pre registered snapshot names, rather than a single
count. Add a snapshot to failure stats if is in pre-registered set
but is not currently running.

parkertimmins committed:
some cleanup

parkertimmins committed:
d

parkertimmins committed:
Use snapshotId instead of snapshotName

Storing snapshot pre-register stats by name does not work because
there can be multilpe snapshots with the same name (but different
id). So, generate uuid before creating snapshot and use this, along
with name, to identify the stats of a single snapshot.

parkertimmins committed:
spotless

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins committed:
Revert code that I didn't mean to commit

This reverts commit 591276de9464afa2af0b08c311572b52aa47eefe.

parkertimmins committed:
Fix broken test

Two issues:
- calling create snapshot api is now async so needs to be checked in assertBusy
- cluster state in test didn't have any nodes, so failed on pre-registration

parkertimmins committed:
more test fixes`

parkertimmins committed:
pre_registered_snapshots field should be optional

parkertimmins commented on x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java:
```diff
@@ -92,82 +99,108 @@ public static Optional<String> maybeTakeSnapshot(
+            CreateSnapshotRequest request = policyMetadata.getPolicy().toRequest(TimeValue.MAX_VALUE, UUIDs.randomBase64UUID());
+            final SnapshotId snapshotId = new SnapshotId(request.snapshot(), request.uuid());
+
+            submitUnbatchedTask(
```
`submitUnbatchedTask` is deprecated and should be replaced with  `ClusterStateTaskExecutor`. I have not done this as I don't think we actually could do any batching here. As we don't want to/can't batch anything it seems better to explicitly used the un-batched form. Or should I follow the deprecation warning and update both calls to use `ClusterStateTaskExecutor`?

parkertimmins committed:
cleanup

parkertimmins commented on x-pack/plugin/slm/src/internalClusterTest/java/org/elasticsearch/xpack/slm/SLMStatDisruptionIT.java:
```diff
@@ -0,0 +1,564 @@
+            assertEquals(0, snapshotLifecyclePolicyMetadata.getInvocationsSinceLastSuccess());
+            assertEquals(Set.of(), snapshotLifecyclePolicyMetadata.getPreRegisteredSnapshots());
+        }, 1, TimeUnit.MINUTES);
+    }
```
I've been unable to build a test that checks the situation where a  snapshot is currently running and another fails. To do this, I need a way to delay the first snapshot so that it is not un-registered until after the second snapshots completes. I haven't figured out a way to do this; any ideas?

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins committed:
add unit tests for PreRegisterSLMRun task

parkertimmins committed:
Fix bug found in unit test!

parkertimmins committed:
unit test for updating preRegister set after snapshot run

parkertimmins committed:
Temp: attempt at making 1 snapshot in policy take longer

parkertimmins committed:
Temp: another save point for working with concurrently running snapshots

parkertimmins committed:
Temp: another save point for working with concurrently running snapshots

parkertimmins committed:
Temp: another save point for working with concurrently running snapshots

parkertimmins committed:
add integration test for concurrently running snapshots

parkertimmins committed:
add locale to String.format usage

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins commented on x-pack/plugin/slm/src/internalClusterTest/java/org/elasticsearch/xpack/slm/SLMStatDisruptionIT.java:
```diff
@@ -0,0 +1,697 @@
+
+        // Use static vars since instantiated by plugin system
+        private static final AtomicBoolean doDelay = new AtomicBoolean(true);
+        private static final CountDownLatch delayedRepoLatch = new CountDownLatch(1);
```
Making these static is a bit silly. It'd be better if they were instance fields within `TestDelayedRepo`. But I couldn't figure out how to get only get a single instance from PlugService causing the version of `delayedRepoLatch` being waited on to be  a separate instance from that which had `countDown` called.

dakrone replied:
I think it's fine that they're static

parkertimmins commented on x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java:
```diff
@@ -204,37 +238,144 @@ public static String exceptionToString(Exception ex) {
+        }
+
+        @Override
+        public void onFailure(Exception e) {
```
Do you think there is a better way to handle this failure? This will be called if there is a failure while attempting to preRegister a snapshot.

parkertimmins committed:
test cleanup

parkertimmins committed:
spotlessApply making questionable formatting decisions

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

dakrone commented on x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java:
```diff
@@ -204,37 +238,144 @@ public static String exceptionToString(Exception ex) {
+        return currentlyRunning;
+    }
+
+    static class PreRegisterSLMRun extends ClusterStateUpdateTask {
```
Can you add a general overview of what this task does as javadoc here?

dakrone commented on x-pack/plugin/slm/src/internalClusterTest/java/org/elasticsearch/xpack/slm/SLMStatDisruptionIT.java:
```diff
@@ -0,0 +1,638 @@
+        assertTrue(latch.await(1, TimeUnit.MINUTES));
+
+        // restart master so failure stat is lost
+        // TODO this relies on a race condition.
+        // The node restart must happen before stats are stored in cluster state, but this is not guaranteed.
```
Are these TODOs still relevant? Is this going to cause failures down the road?

parkertimmins replied:
The race condition is still present ... but I've never actually seen it not fail to lose the stats. I think it's probably fine to remove the TODO, but leave a big comment saying: "if this test fails it might be because of this race condition"

dakrone replied:
Do you mean that you've never seen it lose the stats? Or do you mean when it occurs it always loses the stats? If it hasn't failed then removing the TODOs (I think there are a couple of them) and replacing with the other comment sounds good to me

parkertimmins replied:
Yeah, the intention of that line it to force it to lose stats. So far, it has always succeeded in losing the stats. Cool, then I'll remove the todo.

dakrone reviewed (changes requested):
I left some minor comments, thanks for iterating on this! I like this approach better than the two-CS-updates-in-parallel one.

parkertimmins committed:
Set CreateSnapshotRequest uuid in constructor

parkertimmins commented on x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java:
```diff
@@ -204,37 +238,144 @@ public static String exceptionToString(Exception ex) {
+            SnapshotLifecycleStats stats = snapMeta.getStats();
+            long unrecordedFailures = 0;
+            Set<SnapshotId> newPreRegisteredSnapshots = new HashSet<>();
+            for (SnapshotId snapshot : preRegisteredSnapshots) {
```
Good point, yep that makes sense :+1:

parkertimmins committed:
Change preRegistered stat from set to list

parkertimmins committed:
make it spotless!

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins committed:
change warn to debug

parkertimmins commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/CreateSnapshotRequest.java:
```diff
@@ -81,6 +81,7 @@ public class CreateSnapshotRequest extends MasterNodeRequest<CreateSnapshotReque
 
     @Nullable
     private Map<String, Object> userMetadata;
+    private String uuid = null;
```
I'm having a bit of trouble with this one. 

It makes sense that `source` and `toXContent` should be consistent, but it's seeming a bit weird to parse `uuid` within source. It seems this would allow users to specify a `uuid` within their SLM policy definition config.

On the flipside, now I'm wondering if `uuid` should be in CreateSnapshotRequest.toXContent? When is a snapshot request even converted to xcontent? I don't ~think~ it's stored in the cluster state or return it via an api endpoint.

parkertimmins replied:
It appears that `toXContent()` is only used to test `source`. Since `source` shouldn't parse uuid, it could be removed from toXContent. Alternatively, it could just be in toXContent but not in source. This is how the two other fields set in the constructor, `repository` and `snapshot`, are treated. But they could also just be removed from x content, since it's only used for testing. What do you think?

dakrone replied:
Sounds fine to me to remove it from `toXContent`

parkertimmins committed:
Remove uuid from CreateSnapshotRequest x content

parkertimmins committed:
Fix broken test, remove TODOs

parkertimmins committed:
Set uuid in CreateSnapshotRequest if not set in constructor

parkertimmins commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/CreateSnapshotRequest.java:
```diff
@@ -364,6 +373,23 @@ public CreateSnapshotRequest userMetadata(@Nullable Map<String, Object> userMeta
+    }
+
+    public String uuid() {
+        // If request was constructed on node with transport version < PRE_REGISTER_SLM_STATS it will not have a uuid yet
```
Set the uuid in the getter to deal with the rare case that the node that created the request had a transport version before PRE_REGISTER_SLM_STATS and thus didn't set uuid.

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins committed:
Fix test merged from main

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins committed:
Fix remaining issue from merge with main

parkertimmins committed:
register snapshot within snapshotservice

parkertimmins committed:
partial work using map from policy to snapshots

parkertimmins committed:
Before deleting starting task

parkertimmins committed:
Update registered after snapshot only

parkertimmins committed:
add some unit tests

parkertimmins committed:
Replace map with flat list of policy/snapshotId

parkertimmins committed:
Change preRegister to register

parkertimmins committed:
Move policy logic out of snapshot service

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins committed:
A bit more cleanup of tests

parkertimmins committed:
spotlessApply

parkertimmins committed:
change another instance of pre-registered to registered

parkertimmins committed:
spotlessApply

parkertimmins committed:
spotless

parkertimmins committed:
Add registered policy snapshots serialization tests

parkertimmins committed:
more spotlessApply :(

dakrone reviewed (approved):
This LGTM, I left only a few minor comments, thanks for iterating on this Parker!

parkertimmins committed:
Add java docs, cleanup

parkertimmins commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/CreateSnapshotRequest.java:
```diff
@@ -364,6 +373,23 @@ public CreateSnapshotRequest userMetadata(@Nullable Map<String, Object> userMeta
+        return this;
+    }
+
+    public String uuid() {
+        // If request was constructed on node with transport version < REGISTER_SLM_STATS it will not have a uuid yet
+        if (this.uuid == null) {
+            this.uuid = UUIDs.randomBase64UUID();
+        }
+        return this.uuid;
+    }
```
I don't think there are multiple threads on the same node using this. Thus, though it would be easy to make threadsafe, I'm inclined to mark it as non-threadsafe so it doesn't give the impression that it needs to be threadsafe.

I've been thinking about this trying to come up with scenarios that could behave incorrectly. 
I'm not sure if this is really possible but I imagine a scenario where two uuid's might be generated for the same request.  Say there are 3 nodes with versions: 9, 7, 9.
- the request is constructed on a node with version 9 and thus has a uuid generated
- transferred through node with version 7, thus loses uuid
- lastly sent to the other node with version 9, which generates a new uuid.

Do you think it would be an option to fail requests in `validate()` if uuid is null? This would cause failure on snapshots requests during upgrades, but seems safer.

dakrone replied:
I don't think there are multiple threads using this either. Marking it is fine. I also don't think that the 9->7->9 scenario is going to happen either, since the master node is the one doing these things, and not transferring it both and forth between nodes. So I think documenting it is a fine solution.

parkertimmins replied:
I am a bit confused about `master node is the one doing these things,`. In the case of SnapshotLifecycleTask that makes sense I think as SnapshotLifecycleService is only running on the master node. But for ad-hoc snapshot created through the [rest api](https://github.com/elastic/elasticsearch/blob/5662f988b208718b8b3a1f1a1b0e549ac0a2d563/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestCreateSnapshotAction.java#L46) is this true? (Although I agree my example seems pretty darn contrived)

parkertimmins replied:
I think you are right that this is fine as is, I'm just wanting to understand if something like this is possible.

Also, I'm thinking I'll split this getter into two methods: a getter `uuid()` and `getOrCreateUUID()` as making a "getter" not thread-safe seems to defy expectations.

parkertimmins commented on server/src/test/java/org/elasticsearch/snapshots/RegisteredPolicySnapshotsSerializationTests.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    private RegisteredPolicySnapshots.PolicySnapshot randomPolicySnapshot() {
+        SnapshotId snapshotId = new SnapshotId(randomIdentifier(), randomUUID());
```
It shouldn't be possible in the real code for UUID to be null here. It should serialize fine, but this class is only meant to test valid  objects, right?

dakrone replied:
I think it could be `null` if sent to a newer node from a pre-this-change master node?

parkertimmins replied:
I believe in that case the call to `request.uuid()` at https://github.com/elastic/elasticsearch/pull/110216/files#diff-de99cbad875eb82467666e03b94c5b9376e92ce73fa021a99ab58ea4c2bae4f6R280 should have set the uuid on the request before it is then added to the registered snapshots at https://github.com/elastic/elasticsearch/pull/110216/files#diff-de99cbad875eb82467666e03b94c5b9376e92ce73fa021a99ab58ea4c2bae4f6R3963

Am I not considering a way this could happen?

Also, looks like uuid can't be null in SnapshotId: https://github.com/elastic/elasticsearch/blob/c7819aa91adb41fac547c406d1d1d83de3dae5d4/server/src/main/java/org/elasticsearch/snapshots/SnapshotId.java#L56

parkertimmins committed:
Add javadoc, fix test

parkertimmins committed:
Fix race condition when restart during test

There was a race condition in tests where restart
could be called too early or too late. If too early,
the snapshot before would yet be completed yet. If too
late, SLM would already have succeeeded in writing failure
stats.

parkertimmins committed:
Merge branch 'main' into snapshot-failure-stats

parkertimmins committed:
internal test compilation fix

parkertimmins committed:
one more unit test
